Controllers/ExportController.cs

```diff
@@ -0,0 +1,243 @@
+
+            var query = _session.Query<WorkflowType, WorkflowTypeIndex>()
+                .OrderBy(x => x.Name);
+
+            var count = await query.CountAsync();
+
+            var workflowTypes = await query
+                .Skip(pager.GetStartIndex())
+                .Take(pager.PageSize)
+                .ListAsync();
+            var workflowTypeIds = workflowTypes.Select(x => x.WorkflowTypeId).ToList();
+            var workflowInstances = (await _session.QueryIndex<WorkflowIndex>(x => x.WorkflowTypeId.IsIn(workflowTypeIds))
+                .ListAsync())
+                .GroupBy(x => x.WorkflowTypeId)
+                .ToDictionary(x => x.Key);
+
+            var pagerShape = (await New.Pager(pager)).TotalItemCount(count);
+
+            var model = new WorkflowExportListViewModel
+            {
+                WorkflowTypes = workflowTypes
+                .Select(x => new WorkflowTypeEntry
+                {
+                    WorkflowType = x,
+                    Id = x.Id,
+                    Name = x.Name,
+                    WorkflowCount = workflowInstances.ContainsKey(x.WorkflowTypeId) ? workflowInstances[x.WorkflowTypeId].Count() : 0
+                })
+                .ToList(),
+                Pager = pagerShape
+            };
+            return View(model);
+        }
+
+        #endregion Index
+
+        #region Preview
+
+        public async Task<IActionResult> Preview(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            var workflowType = await _session.GetAsync<WorkflowType>(id.Value);
+
+            if (workflowType == null)
+            {
+                return NotFound();
+            }
+
+            var workflowInstancesQuery = _session.Query<Workflow, WorkflowIndex>(x => x.WorkflowTypeId == workflowType.WorkflowTypeId)
+                .OrderByDescending(x => x.CreatedUtc);
+
+            var instancesCount = await workflowInstancesQuery.CountAsync();
+            var preview = await workflowInstancesQuery.FirstOrDefaultAsync();
+
+            var model = new PreviewWorkflowExportViewModel
+            {
+                Name = workflowType.Name,
+                InstanceCount = instancesCount,
+                PreviewOutput = GetOutput(preview),
+                WorkflowTypeId = id.Value
+            };
+
+            return View(model);
+        }
+
+        #endregion Preview
+
+        #region Export
+
+        [HttpPost]
+        public async Task<IActionResult> Export(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            var workflowType = await _session.GetAsync<WorkflowType>(id.Value);
+
+            if (workflowType == null)
+            {
+                return NotFound();
+            }
+
+            var query = _session.Query<Workflow, WorkflowIndex>(x => x.WorkflowTypeId == workflowType.WorkflowTypeId)
```

Comment:
@lawrencedine Thoughts on moving the responsibility of creating the memory stream for the CSV file generation to a service class? Personally I'm not sure the controller action should be responsible for generating the CSV. Be good to add a try/catch round that too in case there is issues that could then be logged.

Reply:
@peterkeating I've generally err'd on the side of caution when using `MemoryStream` for file downloads because I think `File()` does some magic to make sure the various Streams get disposed correctly. I had considered breaking it out into a service class though and I think I will, update coming...

Agree on logging also 👍